delta-app/src/main/java/io/cdap/delta/store/DBReplicationOffsetStore.java

```diff
@@ -0,0 +1,103 @@
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.delta.store;
+
+import io.cdap.cdap.spi.data.StructuredRow;
+import io.cdap.cdap.spi.data.StructuredTable;
+import io.cdap.cdap.spi.data.StructuredTableContext;
+import io.cdap.cdap.spi.data.TableNotFoundException;
+import io.cdap.cdap.spi.data.table.StructuredTableId;
+import io.cdap.cdap.spi.data.table.StructuredTableSpecification;
+import io.cdap.cdap.spi.data.table.field.Field;
+import io.cdap.cdap.spi.data.table.field.FieldType;
+import io.cdap.cdap.spi.data.table.field.Fields;
+import org.apache.commons.lang.StringUtils;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * A StructuredTable based database table to store replication offset and sequence data.
+ */
+public class DBReplicationOffsetStore {
+
+  private static final StructuredTableId TABLE_ID = new StructuredTableId("delta_offset_store");
+  private static final String NAMESPACE_COL = "namespace";
+  private static final String GENERATION_COL = "generation";
+  private static final String APP_NAME_COL = "app_name";
+  private static final String INSTANCE_ID_COL = "instance_id";
+  private static final String OFFSET_COL = "offset";
+  private static final String UPDATED_COL = "last_updated";
+
+  public static final StructuredTableSpecification TABLE_SPEC = new StructuredTableSpecification.Builder()
+    .withId(TABLE_ID)
+    .withFields(new FieldType(NAMESPACE_COL, FieldType.Type.STRING),
+                new FieldType(GENERATION_COL, FieldType.Type.LONG),
+                new FieldType(APP_NAME_COL, FieldType.Type.STRING),
+                new FieldType(INSTANCE_ID_COL, FieldType.Type.INTEGER),
+                new FieldType(OFFSET_COL, FieldType.Type.STRING),
+                new FieldType(UPDATED_COL, FieldType.Type.LONG))
+    .withPrimaryKeys(NAMESPACE_COL, APP_NAME_COL, GENERATION_COL, INSTANCE_ID_COL)
+    .build();
+
+  private final StructuredTable table;
+
+  private DBReplicationOffsetStore(StructuredTable table) {
+    this.table = table;
+  }
+
+  static DBReplicationOffsetStore get(StructuredTableContext context) {
+    try {
+      StructuredTable table = context.getTable(TABLE_ID);
+      return new DBReplicationOffsetStore(table);
+    } catch (TableNotFoundException e) {
+      throw new IllegalStateException(String.format(
+        "System table '%s' does not exist. Please check your system environment.", TABLE_ID.getName()), e);
+    }
+  }
+
+  public String getOffsets(String namespace, String appName, Long generation, int instanceId)
```

Comment:
did the implementation change? Now it looks like it can't return null.

Reply:
also, this store should be working with the OffsetAndSequence object, not with strings.

Objects should only be encoded/decoded at the very edges -- in the handler or in the store. That way we ensure that we are dealing with proper representations and not malformed strings.

Reply:
The generation also should be a long instead of a Long, since it shouldn't ever be null. This method should be:

`public OffsetAndSequence getOffset(DeltaWorkerId workerId)`